dotCMS/src/main/java/com/dotcms/auth/providers/saml/v1/SAMLHelper.java

```diff
@@ -0,0 +1,469 @@
+
+        if (!DotSamlConstants.DOTCMS_SAML_BUILD_ROLES_NONE_VALUE.equalsIgnoreCase(buildRolesStrategy)) {
+            try {
+                // remove previous roles
+                if (!DotSamlConstants.DOTCMS_SAML_BUILD_ROLES_STATIC_ADD_VALUE.equalsIgnoreCase(buildRolesStrategy)) {
+
+                    Logger.debug(this, ()-> "Removing ALL existing roles from user '" + user.getUserId() + "'...");
+                    this.roleAPI.removeAllRolesFromUser(user);
+                } else {
+
+                    Logger.debug(this, ()-> "The buildRoles strategy is: 'staticadd'. It won't remove any existing dotCMS role");
+                }
+
+                this.handleRoles(user, attributesBean, identityProviderConfiguration, buildRolesStrategy);
+            } catch (DotDataException e) {
+
+                Logger.error(this, "Error adding roles to user '" + user.getUserId() + "': " + e.getMessage(), e);
+                throw new SamlException(e.getMessage());
+            }
+        } else {
+
+            Logger.info(this, ()->"The build roles strategy is 'none'. No user roles were added/changed.");
+        }
+    }
+
+    private void handleRoles(final User user, final Attributes attributesBean,
+                             final IdentityProviderConfiguration identityProviderConfiguration,
+                             final String buildRolesStrategy) throws DotDataException {
+
+        this.addRolesFromIDP(user, attributesBean, identityProviderConfiguration, buildRolesStrategy);
+
+        // Add SAML User role
+        this.addRole(user, DotSamlConstants.DOTCMS_SAML_USER_ROLE, true, true);
+        Logger.debug(this, ()->"Default SAML User role has been assigned");
+
+        // the only strategy that does not include the saml user role is the "idp"
+        if (!DotSamlConstants.DOTCMS_SAML_BUILD_ROLES_IDP_VALUE.equalsIgnoreCase(buildRolesStrategy)) {
+            // Add DOTCMS_SAML_OPTIONAL_USER_ROLE
+            if (DotSamlProxyFactory.getInstance().samlConfigurationService().getConfigAsString(identityProviderConfiguration,
+                    SamlName.DOTCMS_SAML_OPTIONAL_USER_ROLE) != null) {
+
+                this.addRole(user, DotSamlProxyFactory.getInstance().samlConfigurationService().getConfigAsString(identityProviderConfiguration,
+                        SamlName.DOTCMS_SAML_OPTIONAL_USER_ROLE), false, false);
+                Logger.debug(this, ()-> "Optional user role: " +
+                        DotSamlProxyFactory.getInstance().samlConfigurationService().getConfigAsString(identityProviderConfiguration,
+                                SamlName.DOTCMS_SAML_OPTIONAL_USER_ROLE) + " has been assigned");
+            }
+        } else {
+
+            Logger.info(this, "The build roles strategy is 'idp'. No saml_user_role has been added");
+        }
+    }
+
+    private boolean isValidRole(final String role, final String[] rolePatterns) {
```

Comment:
Codacy found an issue: [Consider using varargs for methods or constructors which take an array the last parameter.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5707914)